test.sh

```diff
@@ -21,25 +21,118 @@ TESTDIRS="analysis \
           # cmd
           # Godeps
 
+# We need to know, for github-pr-status, what the triggering commit is.
+# Assume first it's the travis commit (for builds of master), unless we're
+# a PR, when it's actually the first parent.
+TRIGGER_COMMIT=${TRAVIS_COMMIT}
+if [ "${TRAVIS_PULL_REQUEST}" != "false" ] ; then
+  revs=$(git rev-list --parents -n 1 HEAD)
+  # The trigger commit is the last ID in the space-delimited rev-list
+  TRIGGER_COMMIT=${revs##* }
+fi
+
+start_context() {
+  CONTEXT="$1"
+  echo "Starting [${CONTEXT}]"
+}
+
+end_context() {
+  CONTEXT=""
+}
+
+update_status() {
+  if [ "${TRAVIS}" == "true" ] && [ "x${CONTEXT}" != "x" ] ; then
+    github-pr-status --authfile "$(pwd)/test/github-secret.json" \
+      --owner "letsencrypt" --repo "boulder" \
+      status --sha "${TRIGGER_COMMIT}" --context "${CONTEXT}" $*
+  fi
+}
+
 run() {
   echo "$*"
-  if $*; then
+  $* 2>&1
+  local status=$?
+
+  if [ ${status} -eq 0 ]; then
+    update_status --state success
     echo "success: $*"
   else
     FAILURE=1
+    update_status --state failure
     echo "failure: $*"
   fi
+
+  return ${status}
 }
 
+run_and_comment() {
+  if [ "x${TRAVIS}" = "x" ] || [ "${TRAVIS_PULL_REQUEST}" == "false" ] ; then
+    run $*
+  else
+    result=$(run $*)
+    local status=$?
+    # Only send a comment if exit code > 0
+    if [ ${status} -ne 0 ] ; then
+      echo $'```\n'${result}$'\n```' | github-pr-status --authfile "$(pwd)/test/github-secret.json" \
+        --owner "letsencrypt" --repo "boulder" \
+        comment --pr ${TRAVIS_PULL_REQUEST}
+    fi
+  fi
+}
 
 # Path for installed go package binaries. If yours is different, override with
 # GOBIN=/my/path/to/bin ./test.sh
 GOBIN=${GOBIN:-$HOME/gopath/bin}
 
-# Ask vet to check in on things
-run go vet -x ./...
+#
+# Run Go Vet, a correctness-focused static analysis tool
+#
+
+start_context "test/vet"
+run_and_comment go vet ./...
```

Comment:
This is the consequence of calling it end_context instead of clear_context, but I'm good with adding it places - it's not like it's expensive, and I like that it adds clarity.